RFD 98 - Secure OCI images (#19045)
gravitational/teleport: rfd/0098-securing-docker-images.md

---
authors: Trent Clarke (trent@goteleport.com)
state: draft
---

# RFD 0098 - Delivering secure Teleport OCI Images

# Required Approvers
* Engineering: @r0mant
* Security: @reed
* Product: (@xinding33 || @klizhentas)

## What

This RFD discusses structures and processes to increase the security of the OCI 
images we use to deliver Teleport to our customers.

## Why

> One of our shipping channels is a Docker image. As the provider of that docker

Reviewer: Well, technically the Docker image is the _artifact_, and ECR is the shipping channel :-)

> image we are (at least partialy) responsible for everything on it, not just

Reviewer 2: ```suggestion
image we are (at least partialy) responsible for everything in it, not just
```

Teleport. We should not ship vulnerabilities to our clients, even if those 

> vulnerabilities are not odirectly in our software.

Reviewer: ```suggestion
vulnerabilities are not directly in our software.
```

## Details

### Goals

For the sake of this RFD, I will define _delivering a secure image_ as 

> Reliably producing an OCI image that:
>
>
>  1. a priori is unlikely to contain vulnerabilities in and of itself, 
>  2. contains software of known provenance,
>  3. has no warnings or vulnerabilities flagged when run through a reputable 
>     scanner at the time of creation, and
>  4. flags vulnerabilities even after creation, either in Teleport itself or 
>     a dependency.

### Non-goals

This RFD will not discuss building images for and/or running Teleport on
platforms other than Linux.

### Approach

This RFD describes a 2-pronged approach for meeting the above goals:

   1. Switching to distroless base images, and 

>    2. Automated, ongoing monitoring of images using a scanning service

Reviewer 3: 3. Producing new images that resolve any vulnerabilities in a timely fashion to meet our vulnerability management policy (maximum length of time for fixes -- critical - 7 days, high - 30 days, moderate - 90 days, low - no set timeline (but preferably 180 days)).

### 1. Distroless Images

> Distroless images contain only an application and the minimal set of

Reviewer: Do all of our features work in distroless images? I'm thinking things like PAM, BPF, Rust (GLIBC dependency), etc.

Reviewer 4: We don't enable SSH service in any of our Helm charts so I'm not sure getting things like PAM and BPF working is needed tbh. Not sure if any of our customers enable SSH service in their custom configs, or what a use-case would even be since we have Kube access.

Couple of items that we might need to think about though are:

- Desktop access is also currently not present at all in Helm charts. Do we need/want to support it? If so, we might need to make sure Rust stuff works, yeah.
- We will likely be looking at adding BPF-like advanced session recording to Kubernetes access relatively soon. No idea how it would work yet but I wonder if moving to distroless now is going to complicate things somewhat.

Author: I've picked the CC image specifically for GLIBC support, which is all our Rust code should need. 

As far as PAM and BPF support is concerned, my understanding is that we only nee specific kernel versions to make this work, and we link `libpam` etc statically. It's definitely worth some practical testing, though.

Author: And practical tests confirm that I'm wrong, at least for PAM. Looking into adding PAM support

Author: We might be able to go back to the distroless source and rebuild our image from there - after all, if _they_ can add arbitrary debian packages to their images then we should be able to as well. 

I'm trying to understand `bazel` enough to see if this is a reasonable option. (The default GHA builder image has `bazel` already installed, so that's a start)

Author: Answer: You _can_ add arbitrary Debian packages to an image using `bazel`, but the method used by `rules_docker` and `distroless` is functionally no different from the unpack-and-copy method used in the Dockerfile, so there doesn;t seem to be much point in making the tooling more complicated.

dependencies for it. Google offers several base images that contain minimal 
Linux distribution that we can use as a starting point.

Switching to distroless images drastically reduces the number of software components 
we ship as part of a Teleport distribution. This both reduces the size of the potential
attack surface, and reduces the potential for high-noise reports from automated
scanners. 

The Google-supplied distroless base images also [provide a mechanism for verifying the
provenance of a given image](https://github.com/GoogleContainerTools/distroless#how-do-i-verify-distroless-images) 
using `cosign` and a public key. Stronger, SLSA-2 level guarantees [can be verified with additional
tooling](https://security.googleblog.com/2021/09/distroless-builds-are-now-slsa-2.html). 

**Note** that we are _already_  using Distroless images to distribute some Teleport 
plugins. This would extend their use to Teleport proper.

### 2. Ongoing scanning

Using an _ongoing_ automated scanner means that we do not just check for 
vulnerabilities at image creation time, instead we _continually_ scan for any vulnerabilities 

> that may be discovered during the image's lifetime.

Reviewer 3: What do we define the lifetime as?

Author: For the sake of argument I am defining it as:

> _lifetime_ is the period of time from publishing the image until until either:
> * the image is replaced with a newer verion of the image, or
> * the support lifetime for the verion of Teleport on that  image expires (i.e. falls
  out of our 3-version support window)

## Implementation Details

## 1. Image construction

### Teleport Image Requirements

What is the minimal set of requirements to run Teleport on Linux in a container?

   1. Teleport
   2. `GLIBC` >= 2.17
   3. `dumb-init` is required for correct signal and child processes handling
      inside a container.
   4. CA certificates

Requirement (1) (i.e. Teleport itself) is provided by our CI process. 

Requirements (2) and (4) are satisfied automatically by using the the google-
provided base image [`gcr.io/distroless/cc-debian11`](https://github.com/GoogleContainerTools/distroless#what-images-are-available), which is configured for "mostly statically compiled" languages that require libc.

Requirement (3) (`dumb-init`) can be sourced either from the upstream Debian repository, or
downloaded directly from the project's GitHub release page. Sourcing `dumb-init` from the 
Ubuntu or Debian package repositories implies some minimal provenance checking by the debian
packaging tools, so we will prefer that to sourcing it from GitHub.

> ### Base image verification

Author: Hadn't thought of that :facepalm:, but that is an excellent idea.

> ### Base image verification

Reviewer 5: Could you please this do RFD

fixes, etc. included) while still validating the provenance of the base image
itself.

It is technically possible for the image to be poisoned post-validation (e.g.
in a shared build environment, where a malicious peer could re-tag a malicious
image as the base).

While we _could_ verify that the validated base image appears in the parent 
chain of the final image, this is still no protection against the malicious 
image being based on the same parent. We could _also_ assert that the final 
image's parent chain has an expecetd number of steps (inferred from the number
of steps in the Docker file), but this is erorr-prone and would drastically 
reduce the flexibility of the build system.

For the above reasons, Teleport images for public consumption must not be 
built in such a shared environment.

### Building the image

The image will be built from a multi-stage docker file, using build stages to download
and unpack the required debian packages and copy them into place on the distroless 
image. 

An example Dockerfile, assuming the Teleport Debian package is supplied by the 
CI system, might look like something like:

```Docker

> FROM debian:11 as dumb-init

Reviewer 3: How will we ensure we're staying on top of any security updates to `dumb-init` or any other one-off packages we install?

RUN apt update && apt-get download dumb-init && dpkg-deb -R dumb-init*.deb /opt/dumb-init

FROM debian:11 as teleport
COPY teleport*.deb
RUN dpkg-deb -R teleport*.deb /opt/teleport

FROM gcr.io/distroless/cc-debian11
COPY --from=dumb-init /opt/dumb-init/bin/dumb-init /bin
COPY --from=teleport /opt/teleport/bin/* /bin
ENTRYPOINT ["/bin/dumb-init", "teleport", "start", "-c", "/etc/teleport/teleport.yaml"]
```

### Troubleshooting Images

Troubleshooting a distroless image is hard, as there are no tools baked into
the image to aid in debugging a deployment.

> If we need to add tooling in order to aid troubleshooting, I propose that we

Reviewer 6: The debug image does start to loose some of the distroless value.  The primary advantage to distroless is the small size and thus reduced footprint for CVE's and maintenance.  But I understand that sometimes it's just easier to have a shell and other tools, it just has to be weighed.

> If we need to add tooling in order to aid troubleshooting, I propose that we

Author: Yep, Updated to mention existing distroless `debug` versions

add a `teleport-debug` image, that builds on the distroless image to include
things like a shell, tools from `busybox`, and so on. I would also suggest 
that, while we should take as much care as possible when constructing and 
monitoring this image, use of the debug image is considered "at your own risk".

> ### Compatability Guarantees

Reviewer 4: ```suggestion
### Compatibility Guarantees
```

We have clients relying on the existing behaviour (and contents) of our images. We

> should treat releaseing these distroless images as a compatability break, and make 

Reviewer 4: ```suggestion
should treat releasing these distroless images as a compatibility break, and make 
```

our customers aware of our intentions well in advance so that they can prepare.

## 2. Scanning and monitoring the image

> There are many options for scanning and monitoring, but given we are already using the

Author: Will investigate. Thanks for the tip.

Amazon ECR, it seems most logical to use the built-in ECR scanning tools to detect
known vulnerabilities in the final images.

We are, in fact, already doing this on image upload, although I don't know if
the signal from the scan is monitored in any way.

> I propose that we use the ECR Advanced Scanning tools, that can be set to run at

Reviewer: So what happens if the scanner detects a vulnerability in the image but not in Teleport? I assume if the vulnerability has a high severity we would want to ship an update. Would we rebuild with the latest base image and the same version of Teleport, or would we need to cut a Teleport release to do this?

Reviewer 2: Note: Amazon ECR only supports immutable tags on the internal registry and not `public.ecr.aws`. This means you could create a tag based off some sort of hash for the internal registry and then update the public image.

> I propose that we use the ECR Advanced Scanning tools, that can be set to run at

Reviewer 4: How does this compare with other image scanning tools like Snyk? It looks like Snyk has support for [Github Actions](https://github.com/snyk/actions) which we would in theory integrate nicely with our CD, for example it could run on cron schedule and automatically update images if needed.

Reviewer 6: Snyk is definitely a decent solution, but I was thinking ECR scanning would be a better initial fit for us.  The primary reason for that being we already have some ECR scanning and alerting in place that we can use as a starting place here.

Reviewer 6: Adding Trivy sounds like a good option to me.  But it also will need a built image before it can be scanned

---
authors: Trent Clarke (trent@goteleport.com)
state: draft
---

# RFD 0098 - Delivering secure Teleport OCI Images

# Required Approvers
* Engineering: @r0mant
* Security: @reedloden
* Product: (@xinding33 || @klizhentas)

## What

This RFD discusses structures and processes to increase the security of the OCI 
images we use to deliver Teleport to our customers.

## Why

> One of our shipping atrifacts is are a collection of OCI images. As the

Reviewer 3: ```suggestion
One of our shipping artifacts is are a collection of OCI images. As the
```

> provider of those OCI images we are (at least partialy) responsible for

Reviewer 3: ```suggestions
provider of those OCI images we are (at least partially) responsible for
```

everything in it, not just Teleport. 

We should not ship vulnerabilities to our clients, even if those 
vulnerabilities are not directly in our software.

## Details

### Goals

For the sake of this RFD, I will define _delivering a secure image_ as

> > Reliably producing an OCI image that is _a priori_ is unlikely to contain 

Reviewer 3: ```suggestion
> Reliably producing an OCI image that _a priori_ is unlikely to contain 
```

> vulnerabilities in and of itself, in that it:
>
>  1. has the smallest footprint reasonably possible,
>  2. contains software of known provenance,
>  3. has no warnings or vulnerabilities flagged when run through a reputable 
>     scanner at the time of creation,
>  4. flags vulnerabilities even after creation, either in Teleport itself or 
>     a dependency, and
>  5. is updated to resolve any discovered vulnerabilities in a timely fashion.

For the purposes of this discussion, a "timely fashion" for updates is as the

> Teleport Vulberability policy:

Reviewer 3: ```suggestion
Teleport Vulnerability Management Policy:
````

>    | Level    | Response time |

Reviewer 3: ```suggestion
   | Severity | Resolution time |
```

   |----------|---------------|
   | Critical | 7 days        |
   | High     | 30 days       |
   | Moderate | 90 days       |

>    | Low:     | ~180 days     |

Reviewer 3: ```suggestion
   | Low      | ~180 days     |
```

Reviewer 3: To be fair, no set time on low severity issues, but would be nice if fixed within 180 days.

   building minimalist, reproducible Alpine linux images, using a declarative 
   format. While I found the tool verly neat, the images it generates are still closer 
   to a "debug" Distroless image. 

   Using `apko` would also require us to build an Alpine linux package for Teleport 
   to integrate it nto the build.

   I seriously considered recommending `apko`, as it has some neat features (e.g.
   automatically producting a SBOM as part of the construction process), but in the 
   end I rejected it because of the extra software included in the resulting images.

### Image signing

In order to allow our customers to validate our published Teleport images, 
our images will be signed using the `cosign` tool, similarly to the Distroless 
base images.

The `cosign` tool [integrates well with GHA](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/), 
and is even included in the template "how to publish a docker image" example workflow. 

We will need to either publish a public key somewhere, or use ["keyless" signing](https://docs.sigstore.dev/cosign/sign/#keyless-signing), 

> which requires the signer authenticating against OICD.

Reviewer 3: ```suggestion
which requires the signer authenticating against OIDC.
```

More information keyless signing:
 * https://docs.sigstore.dev/cosign/sign/#keyless-signing
 * https://docs.google.com/document/d/1461lQUoVqbhCve7PuKNf-2_NfpjrzBG5tFohuMVTuK4/edit
 * https://www.appvia.io/blog/tutorial-keyless-sign-and-verify-your-container-images/#oidc-flow-with-interaction-free-with-github-actions

> > **AUTHOR'S NOTE:** _I'm a bit torn here. Using a simple keypair is pretty

Reviewer 3: Keyless signing seems like the right approach here.
